aten/src/ATen/native/mps/OperationUtils.h

```diff
@@ -223,23 +223,31 @@ struct MPSGraphCache
   MPSCachedGraph* CreateCachedGraph(const std::string& key, CreateCachedGraphBlock createCacheBlock) {
 
     __block MPSCachedGraph* cachedGraph = nil;
+    __block std::optional<c10::TypeError> type_err;
 
     MPSCacheKey hash = std::hash<std::string>{}(key);
 
     dispatch_sync(serialQueue_, ^() {
 
-      // verify the cached entry doesn't already exist
-      if (cache_.count(hash) != 0) {
-        auto& entry = cache_.at(hash);
-        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(key == entry.key_, "Key collision in the MPS cached graph!\n");
-        cachedGraph = entry.cachedGraph_;
-      } else {
-        cachedGraph = createCacheBlock();
-        CacheEntry entry(key, cachedGraph);
-        cache_.emplace(hash, entry);
-        profileCachedGraph(entry);
+      try {
+        // verify the cached entry doesn't already exist
+        if (cache_.count(hash) != 0) {
+          auto& entry = cache_.at(hash);
+          TORCH_INTERNAL_ASSERT_DEBUG_ONLY(key == entry.key_, "Key collision in the MPS cached graph!\n");
+          cachedGraph = entry.cachedGraph_;
+        } else {
+          cachedGraph = createCacheBlock();
+          CacheEntry entry(key, cachedGraph);
+          cache_.emplace(hash, entry);
+          profileCachedGraph(entry);
+        }
+      } catch (const c10::TypeError &err) {
```

Comment:
In one of the tests, `torch.rand` is called for MPS device with complex tensor, which is yet not supported, and this is why type error is catching here. (Otherwise exceptions raised from dispatch block crash the app, and moving all the check outside of the block is not reliable)

Reply:
Actually, learned of a neat C++11 feature: https://en.cppreference.com/w/cpp/error/current_exception